torch/_inductor/codegen/cpp_micro_gemm.py

```diff
@@ -513,12 +513,27 @@ class CppMicroGemmAMX(CppMicroGemm):
     {{kernel.assert_function}}(K % 2 == 0, "K dimension must be multiple of 2");
 {%- if use_cached_dequantized_B %}
     // Create a stack-allocated buffer for tiles of B.
-    // Except maybe for the tail-case, an AMX tile of B has 16x32 BF16 elements,
-    // but to make things simple, we'll allocate space for a 16x32 tile even for the tail.
-    // It's possible that N may not be equal to Nr, but might be a multiple of it.
-    // So, we'll allocate the buffer accordingly.
+    // Except maybe for the tail-case, an AMX tile of B has 16x32 BF16 elements.
     const auto num_elements_per_b_tile = 512;
-    const auto buf_size_per_nr_block = ((K + {{block_k}} - 1) / {{block_k}}) * num_elements_per_b_tile * 2;
+    const auto last_k_offset = K / {{block_k}} * {{block_k}};
+    const auto tail_k_size = K - last_k_offset;
+    const auto buf_size_per_nr_block = ((K / {{block_k}}) * num_elements_per_b_tile + tail_k_size * 32) * 2;
+    // Although the outer loop in the micro-kernel iterates on n, we could only cache K * Nr elements of
+    // dequantized B, pertaining to each value of n.
+    // But we allow for caching of multiple K * Nr sets in the code to support N != Nr for two reasons:
+    // 1. If N would not be equal to Nr in the future, outer loop in the micro-kernel that iterates over n
+    //    could potentially be unrolled. If the cache could only store K * Nr elements, loop-unrolling would be
+    //    impossible due to memory aliasing.
+    // 2. In the future, if the loops that iterate over m & n would need to be inverted,
+    //    such that m would be iterated over in the outer-loop instead, then for correctness,
+    //    we would need multiple dequantized B caches of K * Nr elements, if N would not be equal to Nr,
+    //    but would instead be a multiple of it.
+    // However, if performance would be better with iterating over n in the outer-loop,
+    // then we may revisit this decision because in that case, we could choose K & N such that
+    // K * Nr elements would fit in L1D but K * N elements need not (Other aspects of blocking
```

Comment:
> Since we always design things to make the smaller "K*Nr" fit in L1D shared by A matrix, making N in the outer-loop should always be optimal

Oh, thanks! I was confused about whether it may be possible to choose some values of `M`, `K`, and `N` (at the granularity of the micro-kernel, after adding another level of hierarchical caching) to fit both `Mr * K` and `K * N` elements in L1D in the future, so that if `m` was iterated over in the outer loop, then some A tiles could reuse multiple `K * Nr` weight sets (given `N` would be a multiple of `Nr`).

Currently, we ensure that only `K * Nr` would fit into L1D, and since `n` is in the outer-loop, if `N != Nr` in the future, 
then even if `Mr * K` elements of A could fit alongside `K * Nr` B elements in L1D, the corresponding A tiles would have to be loaded multiple times into L1D, pertaining to each value of `n`.

However, I'm not sure of the feasibility of such an approach, so I'll remove the notes, and retain `n` in the outer-loop.

Thanks!